src/Microsoft.ML.Tokenizers/Model/BertTokenizer.cs

```diff
@@ -0,0 +1,729 @@
+                                            base.EncodeToIds(textSpan, maxTokenCount - 2, out normalizedText, out charsConsumed, considerPreTokenization, considerNormalization) :
+                                            base.EncodeToIds(text, maxTokenCount - 2, out normalizedText, out charsConsumed, considerPreTokenization, considerNormalization);
+
+                if (ids is not List<int> list)
+                {
+                    list = new List<int>(ids);
+                }
+
+                list.Insert(0, ClsTokenId);
+                list.Add(SepTokenId);
+
+                return list;
+            }
+
+            return text is null ?
+                    base.EncodeToIds(textSpan, maxTokenCount, out normalizedText, out charsConsumed, considerPreTokenization, considerNormalization) :
+                    base.EncodeToIds(text, maxTokenCount, out normalizedText, out charsConsumed, considerPreTokenization, considerNormalization);
+        }
+
+        private IReadOnlyList<int> EncodeToIds(string? text, ReadOnlySpan<char> textSpan, bool addSpecialTokens, bool considerPreTokenization = true, bool considerNormalization = true)
+        {
+            IReadOnlyList<int> ids = text is null ? base.EncodeToIds(textSpan, considerPreTokenization, considerNormalization) : base.EncodeToIds(text, considerPreTokenization, considerNormalization);
+
+            if (addSpecialTokens)
+            {
+                if (ids is not List<int> list)
+                {
+                    list = new List<int>(ids);
+                }
+
+                list.Insert(0, ClsTokenId);
+                list.Add(SepTokenId);
+
+                return list;
+            }
+
+            return ids;
+        }
+
+        /// <summary>
+        /// Build model inputs from a sequence or a pair of sequence for sequence classification tasks by concatenating and adding special tokens. A BERT sequence has the following format:
+        ///     - single sequence: `[CLS] tokenIds0 [SEP]`
+        ///     - pair of sequences: `[CLS] tokenIds0 [SEP] tokenIds1 [SEP]`
+        /// </summary>
+        /// <param name="tokenIds0">List of IDs to which the special tokens will be added.</param>
+        /// <param name="tokenIds1">Optional second list of IDs for sequence pairs.</param>
+        /// <returns>The list of IDs with special tokens added.</returns>
+        /// <exception cref="ArgumentNullException">When <paramref name="tokenIds0"/> is null.</exception>
+        public IReadOnlyList<int> BuildInputsWithSpecialTokens(IEnumerable<int> tokenIds0, IEnumerable<int>? tokenIds1 = null)
+        {
+            if (tokenIds0 is null)
+            {
+                throw new ArgumentNullException(nameof(tokenIds0));
+            }
+
+            // Add 2 for [CLS] and [SEP] tokens. Add 1 for [SEP] token if tokenIds1 is not null.
+            int capacity = tokenIds0.Count() + 2 + (tokenIds1 is null ? 0 : tokenIds1.Count() + 1);
```

Comment:
These Count() calls could end up fully enumerating the source IEnumerable, which is problematic for two reasons. First, there's the performance overhead of doing that enumeration, which is likely to swamp any benefits from setting the capacity. But more important, second, there's no guarantee that the second enumeration is going to produce the same sequence as the first enumeration.